src/PostgREST/Query.hs

```diff
@@ -238,37 +234,23 @@ optionalRollback AppConfig{..} ApiRequest{iPreferences=Preferences{..}} = do
 
 -- | Runs local (transaction scoped) GUCs for every request.
 setPgLocals :: AppConfig  -> KM.KeyMap JSON.Value -> BS.ByteString -> [(ByteString, ByteString)] ->
-               ApiRequest -> PgVersion -> DbHandler ()
-setPgLocals AppConfig{..} claims role roleSettings req actualPgVersion = lift $
+               ApiRequest -> DbHandler ()
+setPgLocals AppConfig{..} claims role roleSettings req = lift $
   SQL.statement mempty $ SQL.dynamicallyParameterized
     ("select " <> intercalateSnippet ", " (searchPathSql : roleSql ++ roleSettingsSql ++ claimsSql ++ [methodSql, pathSql] ++ headersSql ++ cookiesSql ++ appSettingsSql))
     HD.noResult configDbPreparedStatements
   where
-    methodSql = setConfigLocal mempty ("request.method", iMethod req)
-    pathSql = setConfigLocal mempty ("request.path", iPath req)
-    headersSql = if usesLegacyGucs
-                   then setConfigLocal "request.header." <$> iHeaders req
-                   else setConfigLocalJson "request.headers" (iHeaders req)
-    cookiesSql = if usesLegacyGucs
-                   then setConfigLocal "request.cookie." <$> iCookies req
-                   else setConfigLocalJson "request.cookies" (iCookies req)
-    claimsSql = if usesLegacyGucs
-                  then setConfigLocal "request.jwt.claim." <$> [(toUtf8 $ K.toText c, toUtf8 $ unquoted v) | (c,v) <- KM.toList claims]
-                  else [setConfigLocal mempty ("request.jwt.claims", LBS.toStrict $ JSON.encode claims)]
-    roleSql = [setConfigLocal mempty ("role", role)]
-    roleSettingsSql = setConfigLocal mempty <$> roleSettings
-    appSettingsSql = setConfigLocal mempty <$> (join bimap toUtf8 <$> configAppSettings)
+    methodSql = setConfigLocalHardcode ("request.method", iMethod req)
+    pathSql = setConfigLocalHardcode ("request.path", iPath req)
+    headersSql = setConfigLocalJson "request.headers" (iHeaders req)
+    cookiesSql = setConfigLocalJson "request.cookies" (iCookies req)
+    claimsSql = [setConfigLocalHardcode ("request.jwt.claims", LBS.toStrict $ JSON.encode claims)]
+    roleSql = [setConfigLocalHardcode ("role", role)]
+    roleSettingsSql = setConfigLocalParam <$> roleSettings
+    appSettingsSql = setConfigLocalParam <$> (join bimap toUtf8 <$> configAppSettings)
```

Comment:
These settings should be safe too as they come from the DBA input.

To be extra safe, I think we can use:

https://github.com/PostgREST/postgrest/blob/3599024e93f6194762c0da158236a5c2dc24d6d2/src/PostgREST/Query/SqlFragment.hs#L158-L166

Maybe let's do this on this PR to close #3019 and add it to the CHANGELOG.

Reply:
Doing more thinking though. This code is on the fast path and seems wasteful to do that escaping.

Instead, how about:

- `roleSettings` comes from the db ([here](https://github.com/PostgREST/postgrest/blob/3599024e93f6194762c0da158236a5c2dc24d6d2/src/PostgREST/Config/Database.hs#L128-L161)). We could validate that they don't contain any `'`.
```sql
-- Note that a postgrest identifier ("name") can contain "'", but we can enforce they don't, this should be rarely needed anyway
select E'aa.bb.cc\'.dd'::name;
     name     
--------------
 aa.bb.cc'.dd

-- we just need to filter them in the query so we don't complicate ourselves with failure handling
```

- For `configAppSettings`, we can validate and fail if they contain a `'`.

Reply:
It would be nice if we could type those settings as a `newtype SafeLiteral = SafeLiteral Text`.

Reply:
Agree. What I suggested seems too complicated and parametrized queries already handle it better.

Then this PR should be enough for solving https://github.com/PostgREST/postgrest/issues/3019 as it is.

Reply:
>  For me, it is too much.

Why? Can you elaborate?

Right now I see all settings names being constants.

Reply:
@wolfgangwalther I don't see any value being hardcoded, only the keys. This is the result of this PR:

```sql
LOG:  execute 12: select set_config('search_path', $1, true), set_config('role', $2, true), set_config('request.jwt.claims', $3, true), set_config('request.method', $4, true), set_config('request.path', $5, true), set_config('request.headers', $6, true), set_config('request.cookies', $7, true)

DETAIL:  parameters: $1 = '"test", "public"', $2 = 'postgrest_test_anonymous', $3 = '{"role":"postgrest_test_anonymous"}', $4 = 'GET', $5 = '/projects'
, $6 = '{"host":"localhost:3000","accept":"*/*","user-agent":"curl/7.81.0"}', $7 = '{}'
```